### ISSUE TITLE
Architecture Incompatibility

### ISSUE THREAD

**Issue by ade-times** [hodgesmr/llm_ai_eo#1]
Hey I was wondering what machine you did this on. I'm on a MAC and keep getting an error regarding the architecture needed to run the model.

> InternalError: Check failed: (lib_handle_ != nullptr) is false: Failed to load dynamic shared library dist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so dlopen(dist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so, 0x0005): tried: 'dist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OSdist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so' (no such file), '/usr/lib/dist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so' (no such file, not in dyld cache), 'dist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64')), '/Users/ademola/Library/Application Support/io.datasette.llm/mlc/dist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/ademola/Library/Application Support/io.datasette.llm/mlc/dist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so' (no such file), '/Users/ademola/Library/Application Support/io.datasette.llm/mlc/dist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64'))

I'm on Intel Mac machine 

I've tried using the `arch -x86_64 /usr/bin/python3` in all my commands to force it to use an x86 arch but no luck

**Comment by hodgesmr** [comment]
This Notebook was written in Google Colab utilizing CUDA.

Per the [MLC installation instructions](https://llm.mlc.ai/docs/install/mlc_llm.html), I would try replacing this first pip install from the Environment section to:

sh
```
!pip install --pre -U -f https://mlc.ai/wheels mlc-chat-nightly mlc-ai-nightly
```

The [documentation of the llm-mlc](https://github.com/simonw/llm-mlc) library may also be helpful.

If that doesn't work, you can try to [fetch the model directly](https://llm.mlc.ai/docs/) ~(I confirmed this on my x86_64 Mac)~:

```sh
git lfs install && mkdir -p dist/prebuilt
git clone https://huggingface.co/mlc-ai/mlc-chat-Llama-2-7b-chat-hf-q4f16_1 \
                          dist/prebuilt/mlc-chat-Llama-2-7b-chat-hf-q4f16_1
```

**Comment by ade-times** [comment]
Thanks for the quick response - I was already following the MLC installation process since it seemed like versions may be a problem.  Others have raised the issue as [well](https://github.com/simonw/llm-mlc/issues/5#issuecomment-1786105825) but not sure I can parse exactly where my problem is. 

But it does not seem to be an issue with the model but instead with this .so file brought in by the llm-mlc package...
`'/Users/ademola/Library/Application Support/io.datasette.llm/mlc/dist/prebuilt/lib/Llama-2-7b-chat-hf-q4f16_1-metal.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' `

Do you know if what that is saying is that the file is built on arm64 or my machine is running arm64. It's quite hard to tell.

When I run this command 
`file /usr/bin/python3`
It returns:

> /usr/bin/python3: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64e:Mach-O 64-bit executable arm64e]
/usr/bin/python3 (for architecture x86_64):	Mach-O 64-bit executable x86_64
/usr/bin/python3 (for architecture arm64e):	Mach-O 64-bit executable arm64e

So seemingly my MAC can run both but not sure..

Did you get it to all run successfully on your x86_64 Mac

**Comment by hodgesmr** [comment]
Indeed, you're correct, I do get that same error when not running on my M1. My mistake! I think the issue you commented on in the llm-mlc repository is the correct place to track this. In the meantime, you can still experiment quite quickly in Colab.

**Comment by ade-times** [comment]
Yea I'll keep an eye out and let you know if it's resolved. Appreciate the quick help
